README
AS400 Nagios Plugin, v 0.21, March 21 2012
----------------------------------------------

********************************************************************************************
*          WARNING                          WARNING                       WARNING          *
*                                                                                          *
* Version 0.21 has been modifire to work with IBM i V7R1 and will most likely not work     *
* on earlier version of IBM i OS.                                                          *
*                                                                                          *
********************************************************************************************


This plugin allows the monitoring of an IBM AS400.  The main goal was to right the plugin in such a way to prevent having to load extra client software on the AS400 or the monitoring server. In order to accomplish this, information is gathered through normal terminal operations.  

Using the telnet service, a connection is established to the AS400 and OS400 commands are executed and the output parsed. Simple, yet effective, and with little over head. 

Current monitoring variables include CPU Load, DB Load, Free Storage, Jobs, Active Jobs, Check if a job is running, Check if a subsystem is running and OUTQ and Message Queue monitoring. 

I have been running this plugin 24x7 for over a year monitoring the mainframe at a local hospital where I work, and have had zero problems. In addition to monitoring the system status of the 400 including the CPU, DB load, storage, 8 jobs, 2 sub systems and the qsysopr messages, I monitor 29 printer queues.

Notes on the OUTQ variable:
---------------------------

Three different variables are used in determing the status of an OUTQ.  First the OUTQ is checked that it has a connected writer, second that it is in RLS status, and third the number of files in the queue are compared with the specified threshold values.  Any or all of these three variables may be disabled through command line switches. As an example I monitor printer queues which always have writers attached to them, (unless the printer is offline), and I also use it to monitor the fax queue, which dosn't have a writer. Also some users store reports in there outqs, so I disable the check on files. 

Notes on the Message Queue variable:
------------------------------------

The Message Queue monitoring presently only warns if there are any unanswered messages in the specified users queue.  For example I use it to monitor qsysopr.

Notes on the Login variable:
----------------------------
I highly recommend the use of this variable in the creation of dependencies for all services you monitor. Especially if the user that you use with the plugin has an expiring password. This will prevent you from getting lots of alerts should the plugin be unable to login. (ie. QINTER is not running, your password expires, etc.) 

Troubleshooting
-------------------------------
As of version .16 debugging ability has been added to the plugin to help you and me solve problems.  If you are having problems with the client either hanging or returning errors, manually run the check from the command line using the debug options, redirect the output to a file and email it to me.  From that I can solve the problem. Do not worry. There is nothing generated that will violate you or your business's privacy, ex. IP addresses, passwords, etc. Below are some examples: 


cd /usr/nagios/libexec
./check_as400 -D -H ashost -v CPU -w 80 -c 90 >debug.log

or

./check_as400 -D -H ashost -v SBS QINTER >debug.log

In the cases where it is hanging, execute the above command wait about 15 seconds for it to process through to where it hangs, and terminate it with a CTRL-C.

Enjoy!

If you have suggestions or would like to submit bugs, visit the main page at github https://github.com/brdude/check_as400/issues or email me at brdude@gmail.com 


Changelog
*********************
 Nagios AS400 Plugin 
*********************

------------------------------------

0.11
-----------
* added checks for errors on login, and for invalid queue name

0.12a
-----------
*fixed tight loop eating all cpu time when not recieving input

*several spots of code cleanups. Mainly places where I wasn't 
checking the return values of functions before continuing.

0.13
-----------
*logout instead of just disconnecting.

0.14
-----------
*Added check for expired password

*Corrected several incorrect exit status commands

*Added check for expiring password

0.15
-----------
*Minor changes to login procedure (to make it more compatible with other 
 version of OS/400)

*Added check for break messages on login

*Added check for command not allowed error

0.16
-----------
*Asthetic change to help output

*Added ability to check if subsystem is running

*Added ability to check number of active jobs in system

*Added ability to check if a job is running

*Added debug functionality (to help you and me solve problems!)

*Modified check for invalid user name

*Modified check for invalid password

*Removed the transmission of 'esc'(asc 27) at beginning of connection

0.17
-----------
*Minor code cleanups

*Added check to see if a login completes.  Use it in dependency definitions
 for about every other service you monitor. If the login fails, it obviously
 cant check anything. All your checks will return error unable to login.
 This will prevent one from getting a zillion alerts in the middle of the night.
 (Not that I would know anything about that!)

*Fixed problem with active jobs not retrieving properly(parse problem).

*Fixed wrksyssts and dspmsg to execute with proper assistance level and
 modified parsing accordingly.
 
To make things language independent:

*Modified the way the variables CPU,DB,US,JOBS,SBS,AJ,CJ are parsed.
 (All but the CJ, MSG and OUTQ variables should work regardless of langauge) 

*Changed the way the "...is allocated..." message is found during logon.
 (To make it work across languages)

*Added a check on the double conversion for cases where a ',' is used
 instead of a '.'

*Modified parsing on CJ variable when job not found

0.18
-----------
*Changed the name of the command vars class(CmdVars.java) to 
 check_as400_cmd_vars.java for standardization

*Created LANG class and substituted on language depended strings
 with constants from LANG class

*Created language definitions for GERMAN

0.18.01 from Shao-Pin, Cheng
-----------
*Added check Expired password
*Added check DB utilization

0.18.02 from Shao-Pin, Cheng
-----------
*Added check QCMN JOB Transaction timeout. (Only check for CUB customize CLP result ) 
*Added debug check logout "Job ending immediately."
*Added check DISK Status.

0.18.03 from Shao-Pin, Cheng
-----------
*Modified the DSPMSG, see the last message and the number of message needing a reply.

0.18.04 from Shao-Pin, Cheng
-----------
*Added check single ASP used. 
*Added check CPUC, when use  Dynamic hardware  resource, CPU load may need consider Current processing capacity .

0.18.05 from Shao-Pin, Cheng
-----------
* Fix some German language definitions, thank Stefan Hlustik

0.19 by Joerg_M based on 0.18
-----------
*spanish Translation for Status 'Active' added
*Termination Token '[0m' in Method 'logout (';53H' only sent in standard SignOn-Screen)
*Implementation of Variable CJS = "Check for Job Status"  

0.20
---------

by Stefan Hlustik

merged 0.19 and 0.18.05.3

*added failsave mechanism to parsing of DSPMSG when message does not match expected format
*changed DBFAULT parsing Offsets 
*changed printUsageDetail so that it shows where -c and -w are compulsory

 0.21
 ---------

 by Rodrigo Menezes

 merged 0.20 and 0.18.06 

 * Modified OUTQ monitoring for V7R1
 * Removed DB utilization as it is no longer present in the WRKSYSSTS screen in V7R1
 * Modified number of jobs monitoring (JOBS) for V7R1
 * Modified disk usage monitoring for V7R1


check_as400_cmd_vars.java
//Nagios Plugin to check an IBM AS/400
//
//Developed June 2003
//Last Modified September 27 2004
//
//This class is used as a varible structure
//holding parsed command line arguments
//

public class check_as400_cmd_vars{
        //These hold the basic information we need to logon
        public String hostName,userName,passWord;

        //These store specific information on a requested operation
        public String outQ,msgUser,subSystem,job,status,chk_status,aspNums,cpuNum,dgDef;

        //These keep track of what we are suppose to be checking!
        public int checkVariable,command,JobFlags,outQFlags=0;

        //these two hold the warn and critical values
        public double tHoldWarn,tHoldCritical;

        //And of course everybody needs debug variables for when
        //things get ugly!
        public boolean DEBUG=false,DEBUG_PLUS=false;
};



langs/check_as400_lang_german.java
//Nagios Plugin to check an IBM AS/400
//
//Developed September 2004 
//Last Modified January 19 2011
//
//This class is used as a varible structure
//holding language conversions for parsing tokens 
//
//Definitions for the GERMAN language
//provided by Boer, Andre, Stefan Hlustik
//

public class check_as400_lang{
	//These constants are referenced during parsing so
	//that the correct phrases are found.

	//This is found at the bottom when you type dspjob (name of a job 
	//that exists)
	public String SELECTION="Auswahl";
	
	//This is found at the end of lists
	public String LIST_END="Ende";

	//Run dspmsg and it will display "No messages available" if there are no 
	//messages
	public String NO_MESSAGES_AVAILABLE="Keine Nachrichten verf";

	//The "password has expired"/"password expires" messages are the messages
	//you get when you login with an account which has an expired/will expire
	//password.
	public String PASSWORD_HAS_EXPIRED="um die Anmeldeanfrage";
	public String PASSWORD_EXPIRES="Vorheriges Anmelden";

	//The "Display Messages" is what you get after logging into an account
	//which displays any messages before continuing to the menu.
	public String DISPLAY_MESSAGES="Display Messages";

	//Run wrkoutq blah* and it will say "(No output queues)"
	public String NO_OUTPUT_QUEUES="Keine Ausgabewarteschlangen";

	//If you type dspsbsd blah it will say "...not found..."
	public String NOT_FOUND="nicht gefunden";
	//If you type dspjob QINTER, it should complain that there are duplicate 
	//jobs and print at the bottom of the window "duplicate jobs found" 
	public String DUPLICATE="Mehrfach vorhandene";

	//if you type dspjob blah, it will respond Job //blah not found
	//Only put the Job // part.
	public String JOB="Job //";

	//If try and execute a command that you are not allowed it will say
	//"library *LIBL not allowed"
	public String LIBRARY_NOT_ALLOWED="Bibliothek *LIBL nicht gefunden";

	//On a login with an expired password we look for "Exit sign-on" on the 
	//screen before we send the F3 to exit and disconnect. 
	public String EXIT_SIGNON="Verlassen";
	
	//If you type WRKACTJOB it may respond "No active jobs to display"
	//when there is no job like searched for in the sytem
	public String NO_JOB_TO_DISPLAY="Keine aktiven Jobs anzuzeigen";
	
	//Messages needing a reply OR Messages not needing a reply   Nachrichten, fuer die eine/keine Antwort erforderlich ist.
	public String MSG_NEED_REPLY="die eine Antwort erf";
	public int MSG_NEED_REPLY_OFFSET=75;
	public String MSG_NOT_NEED_REPLY="die keine Antwort erfor";
	
	//WRKDSKSTS The "Request" message.
	public String REQUEST_WORD="Anford";
	public String DSK_STS_COMPRESSION="Verdichtung";
};


langs/check_as400_lang_english.java
//Nagios Plugin to check an IBM AS/400
//
//Developed September 2004 
//Last Modified January 19 2011
//
//This class is used as a varible structure
//holding language conversions for parsing tokens 
//
//Definitions for the ENGLISH language
//

public class check_as400_lang{
	//These constants are referenced during parsing so
	//that the correct phrases are found.

        //This is found at the bottom when you type dspjob (name of a job 
        //that exists)
	public String SELECTION="Selection";	
	
		//This is found at the end of lists 
    public String LIST_END="Bottom";

        //Run dspmsg and it will display "No messages available" if there are no 
        //messages
	public String NO_MESSAGES_AVAILABLE="No messages available";

        //The "password has expired"/"password expires" messages are the messages
        //you get when you login with an account which has an expired/will expire
        //password.
	public String PASSWORD_HAS_EXPIRED="Password has expired";
	public String PASSWORD_EXPIRES="password expires";

        //The "Display Messages" is what you get after logging into an account
        //which displays any messages before continuing to the menu.
	public String DISPLAY_MESSAGES="Display Messages";

        //Run wrkoutq blah* and it will say "(No output queues)"
	public String NO_OUTPUT_QUEUES="No output queues";

        //If you type dspsbsd blah it will say "...not found..."
	public String NOT_FOUND="not found";

        //If you type dspjob QINTER, it should complain that there are duplicate 
        //jobs and print at the bottom of the window "duplicate jobs found" 
	public String DUPLICATE="Duplicate"; 

        //if you type dspjob blah, it will respond Job //blah not found
        //Only put the Job // part.
	public String JOB="Job //";

        //If try and execute a command that you are not allowed it will say
        //"library *LIBL not allowed"
	public String LIBRARY_NOT_ALLOWED="library *LIBL not allowed";

        //On a login with an expired password we look for "Exit sign-on" on the 
        //screen before we send the F3 to exit and disconnect.
	public String EXIT_SIGNON="Exit sign-on"; 
	
    	//If you type WRKACTJOB it may respond "No active jobs to display"
		//when there is no job like searched for in the sytem
	public String NO_JOB_TO_DISPLAY="No active jobs to display"; 
	
	    //Messages needing a reply OR Messages not needing a reply
	public String MSG_NEED_REPLY="Messages needing a reply";
	public int MSG_NEED_REPLY_OFFSET=75;
	public String MSG_NOT_NEED_REPLY="Messages not needing a reply";

	    //WRKDSKSTS The "Request" message.
	public String REQUEST_WORD="Request";
	public String DSK_STS_COMPRESSION="Compression";
};
